### PR TITLE
Specially handle more scenarios for type parameters with 'allows ref struct' constraint

### PR DESCRIPTION
**Pull request by AlekseyTs**

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@cston, @jjonescz, @dotnet/roslyn-compiler Please review

File: src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs [hunk 1]
```diff
@@ -23232,5 +23218,669 @@ .maxstack  1
...
+    T GetT<T>(Span<int> s) where T : allows ref struct => default;
+}
+", targetFramework: s_targetFrameworkSupportingByRefLikeGenerics);
+
+            comp.VerifyDiagnostics(
+                // (11,15): error CS8352: Cannot use variable 's1' in this context because it may expose referenced variables outside of their declaration scope
+                //         s2 = (s1 ??= s2);
+                Diagnostic(ErrorCode.ERR_EscapeVariable, "s1 ??= s2").WithArguments("s1").WithLocation(11, 15)
+                );
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs:23386]: >} [](http://example.com/codeflow?start=8&length=1)

Consider testing:
```csharp
static T F<T>() where T : allows ref struct
{
    scoped T s1 = default;
    T s2 = default;
    s2 ??= s1;
    return s2;
}
```

File: src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs [hunk 2]
```diff
@@ -23232,5 +23218,669 @@ .maxstack  1
...
+        {
+            return _pT ??= T.Create();
+        }
+    }
+}
+
+public ref struct S1 : I1<S1>
+{
+    public int P {get;set;}
+    public static S1 Create() => default;
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs:23530]: >default [](http://example.com/codeflow?start=33&length=7)

Consider throwing for `S1.Create()` and `S2.Create()`.

File: src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs [hunk 3]
```diff
@@ -23232,5 +23218,669 @@ .maxstack  1
...
+}
+";
+
+            var comp = CreateCompilation(text, targetFramework: s_targetFrameworkSupportingByRefLikeGenerics, options: TestOptions.ReleaseExe);
+
+            var verifier = CompileAndVerify(
+                comp,
+                expectedOutput: ExecutionConditionUtil.IsMonoOrCoreClr ? "123 246" : null,
+                verify: ExecutionConditionUtil.IsMonoOrCoreClr ? Verification.Passes : Verification.Skipped).VerifyDiagnostics();
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs:23659]: >} [](http://example.com/codeflow?start=8&length=1)

Are we testing the corresponding cases for collection expressions?

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs (hunk 3 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs:23659]: > Are we testing the corresponding cases for collection expressions?

`CollectionExpressions_01`?

**Comment by AlekseyTs** [comment]
@jjonescz, @dotnet/roslyn-compiler For the second review

**Review by jjonescz — APPROVED** [review]

**Pull request merged**